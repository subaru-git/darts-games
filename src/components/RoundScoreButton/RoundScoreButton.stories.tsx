import type { Meta, StoryObj } from '@storybook/react';
import RoundScoreButton from './RoundScoreButton';

const meta: Meta<typeof RoundScoreButton> = {
  component: RoundScoreButton,
  title: 'Components/RoundScoreButton',
};

export default meta;
type Story = StoryObj<typeof RoundScoreButton>;
export const Default: Story = { args: { isFinished: false } };
export const Finished: Story = { args: { isFinished: true, disabled: false, result: 'Result' } };
