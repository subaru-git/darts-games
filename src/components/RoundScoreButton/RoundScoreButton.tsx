import React, { FC } from 'react';
import { Button, IconButton, useBreakpointValue, useDisclosure } from '@chakra-ui/react';
import { GiPartyPopper } from 'react-icons/gi';
import { MdOutlineChangeCircle } from 'react-icons/md';
import RoundOverDialog from '../RoundOverDialog';

type RoundScoreButtonProps = {
  isFinished: boolean;
  disabled: boolean;
  result: string;
  onRoundChange: () => void;
  onRoundOver: () => void;
};

const RoundScoreButton: FC<RoundScoreButtonProps> = ({
  isFinished,
  disabled,
  result,
  onRoundChange,
  onRoundOver,
}) => {
  const isMd = useBreakpointValue({ base: false, md: true });
  return (
    <>
      {!isMd ? (
        <MobileRoundScoreButton
          isFinished={isFinished}
          disabled={disabled}
          onRoundChange={onRoundChange}
          onRoundOver={onRoundOver}
          result={result}
        />
      ) : (
        <DesktopRoundScoreButton
          isFinished={isFinished}
          disabled={disabled}
          onRoundChange={onRoundChange}
          onRoundOver={onRoundOver}
          result={result}
        />
      )}
    </>
  );
};

const DesktopRoundScoreButton: FC<RoundScoreButtonProps> = ({
  isFinished,
  disabled,
  result,
  onRoundChange,
  onRoundOver,
}) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  return (
    <>
      {!isFinished ? (
        <Button
          colorScheme={'green'}
          width={'fit-content'}
          onClick={() => onRoundChange()}
          disabled={disabled}
          aria-label={'round change'}
        >
          Round Change
        </Button>
      ) : (
        <>
          <Button
            colorScheme={'orange'}
            width={'fit-content'}
            onClick={() => onOpen()}
            aria-label={'round over'}
          >
            Round Over
          </Button>
          <RoundOverDialog
            isOpen={isOpen}
            onClose={onClose}
            onNewGame={() => {
              onRoundOver();
              onClose();
            }}
            result={result}
          />
        </>
      )}
    </>
  );
};

const MobileRoundScoreButton: FC<RoundScoreButtonProps> = ({
  isFinished,
  disabled,
  result,
  onRoundChange,
  onRoundOver,
}) => {
  const { isOpen, onOpen, onClose } = useDisclosure();
  return (
    <>
      {!isFinished ? (
        <IconButton
          icon={<MdOutlineChangeCircle />}
          colorScheme={'green'}
          size={'sm'}
          width={'auto'}
          onClick={() => onRoundChange()}
          disabled={disabled}
          aria-label={'round change'}
        />
      ) : (
        <>
          <IconButton
            icon={<GiPartyPopper />}
            colorScheme={'orange'}
            size={'sm'}
            width={'auto'}
            onClick={() => onOpen()}
            aria-label={'round over'}
          />
          <RoundOverDialog
            isOpen={isOpen}
            onClose={onClose}
            onNewGame={() => {
              onRoundOver();
              onClose();
            }}
            result={result}
          />
        </>
      )}
    </>
  );
};

export default RoundScoreButton;
