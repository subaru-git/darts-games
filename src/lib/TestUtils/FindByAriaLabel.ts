export const findByAriaLabel = (elements: HTMLElement[], label: string) => {
  return (
    elements.find((button) => button.attributes.getNamedItem('aria-label')?.value === label) ??
    new HTMLElement()
  );
};
