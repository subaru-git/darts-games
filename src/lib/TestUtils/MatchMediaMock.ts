import mediaQuery from 'css-mediaquery'

export const createMatchMedia =
  (width: number) =>
  (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })
