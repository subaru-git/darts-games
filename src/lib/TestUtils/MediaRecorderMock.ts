export const MediaRecorderMock = {
  start: jest.fn(),
  ondataavailable: jest.fn(),
  onerror: jest.fn(),
  state: '',
  stop: jest.fn(),
};
