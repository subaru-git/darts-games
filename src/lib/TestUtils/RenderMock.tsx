import { ReactNode } from 'react';
import { ChakraProvider } from '@chakra-ui/react';
import { render as defaultRender } from '@testing-library/react';
import Providers from '@/contexts/Providers';

const render = (children: ReactNode | ReactNode[]) =>
  defaultRender(
    <ChakraProvider>
      <Providers>{children}</Providers>
    </ChakraProvider>,
  );
export { render };
