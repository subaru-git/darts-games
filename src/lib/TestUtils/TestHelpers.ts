import { waitFor as originalWaitFor, waitForOptions } from '@testing-library/react';

const defaultTimeout = process.env.WAIT_FOR_DEFAULT_TIMEOUT ? parseInt(process.env.WAIT_FOR_DEFAULT_TIMEOUT, 10) : 5000;

export async function waitFor(callback: () => void,  options?: waitForOptions): Promise<void> {
  return originalWaitFor(callback, { timeout: defaultTimeout, ...options });
}