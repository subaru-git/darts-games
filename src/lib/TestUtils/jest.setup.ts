import mocksdk from 'firebase-mock';
import failOnConsole from 'jest-fail-on-console';
import { createMatchMedia } from '@/lib/TestUtils/MatchMediaMock';
import { MediaRecorderMock } from '@/lib/TestUtils/MediaRecorderMock';

jest.mock('next/router', () => ({ useRouter: jest.fn().mockReturnValue({ locale: 'en' }) }));
jest.mock('react-camera-pro');
jest.mock('@/firebase/client', () => ({
  analytics: mocksdk.analytics,
  db: mocksdk.firestore,
  storage: mocksdk.storage,
  auth: mocksdk.auth,
  functions: mocksdk.functions,
}));
jest.mock('@firebase/auth');
beforeAll(() => {
  window.matchMedia = createMatchMedia(1100);
  window.MediaRecorder = (jest.fn() as any).mockImplementation(() => MediaRecorderMock);
  global.ResizeObserver = require('resize-observer-polyfill');
});

failOnConsole();
