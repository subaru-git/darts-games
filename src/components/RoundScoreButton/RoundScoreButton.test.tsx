import { ChakraProvider } from '@chakra-ui/react';
import { render } from '@testing-library/react';
import { createMatchMedia } from '../../lib/TestUtils/MatchMediaMock';
import RoundScoreButton from './RoundScoreButton';

test('should rendering', () => {
  const onRoundChangeMock = jest.fn();
  const onRoundOverMock = jest.fn();
  window.matchMedia = createMatchMedia(1000);
  const { container } = render(
    <ChakraProvider>
      <RoundScoreButton isFinished={false} result={''} onRoundChange={onRoundChangeMock} onRoundOver={onRoundOverMock} disabled={false} />
    </ChakraProvider>,
  );
  expect(container).toMatchSnapshot();
});
