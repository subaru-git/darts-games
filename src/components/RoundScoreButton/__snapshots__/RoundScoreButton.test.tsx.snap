// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should rendering 1`] = `
<div>
  <button
    aria-label="round change"
    class="chakra-button css-bnxwpm"
    type="button"
  >
    Round Change
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
